File: googlepay-base/src/main/java/com/adyen/checkout/googlepay/model/BillingAddressParameters.java
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 18/7/2019.
 */

package com.adyen.checkout.googlepay.model;

import android.os.Parcel;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.adyen.checkout.core.exception.ModelSerializationException;
import com.adyen.checkout.core.model.JsonUtils;
import com.adyen.checkout.core.model.ModelObject;

import org.json.JSONException;
import org.json.JSONObject;

@SuppressWarnings("MemberName")
public class BillingAddressParameters extends ModelObject {

    @NonNull
    public static final Creator<BillingAddressParameters> CREATOR = new Creator<>(BillingAddressParameters.class);

    private static final String FORMAT = "format";
    private static final String PHONE_NUMBER_REQUIRED = "phoneNumberRequired";

    @NonNull
    public static final Serializer<BillingAddressParameters> SERIALIZER = new Serializer<BillingAddressParameters>() {

        @NonNull
        @Override
        public JSONObject serialize(@NonNull BillingAddressParameters modelObject) {
            final JSONObject jsonObject = new JSONObject();
            try {
                jsonObject.putOpt(FORMAT, modelObject.getFormat());
                jsonObject.putOpt(PHONE_NUMBER_REQUIRED, modelObject.isPhoneNumberRequired());
            } catch (JSONException e) {
                throw new ModelSerializationException(BillingAddressParameters.class, e);
            }
            return jsonObject;
        }

        @NonNull
        @Override
        public BillingAddressParameters deserialize(@NonNull JSONObject jsonObject) {
            final BillingAddressParameters billingAddressParameters = new BillingAddressParameters();
            billingAddressParameters.setFormat(jsonObject.optString(FORMAT, null));
            billingAddressParameters.setPhoneNumberRequired(jsonObject.optBoolean(PHONE_NUMBER_REQUIRED));
            return billingAddressParameters;
        }
    };

    private String format;
    private boolean phoneNumberRequired;

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        JsonUtils.writeToParcel(dest, SERIALIZER.serialize(this));
    }

    @Nullable
    public String getFormat() {
        return format;
    }

    public void setFormat(@Nullable String format) {
        this.format = format;
    }

    public boolean isPhoneNumberRequired() {
        return phoneNumberRequired;
    }

    public void setPhoneNumberRequired(boolean phoneNumberRequired) {
        this.phoneNumberRequired = phoneNumberRequired;
    }
}

File: googlepay/src/main/java/com/adyen/checkout/googlepay/model/BillingAddressParameters.kt
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 18/7/2019.
 */
package com.adyen.checkout.googlepay.model

import android.os.Parcel
import com.adyen.checkout.core.exception.ModelSerializationException
import com.adyen.checkout.core.model.JsonUtils.writeToParcel
import com.adyen.checkout.core.model.ModelObject
import com.adyen.checkout.core.model.getStringOrNull
import org.json.JSONException
import org.json.JSONObject

data class BillingAddressParameters(
    var format: String? = null,
    var isPhoneNumberRequired: Boolean = false,
) : ModelObject() {

    override fun writeToParcel(dest: Parcel, flags: Int) {
        writeToParcel(dest, SERIALIZER.serialize(this))
    }

    companion object {
        private const val FORMAT = "format"
        private const val PHONE_NUMBER_REQUIRED = "phoneNumberRequired"

        @JvmField
        val CREATOR = Creator(BillingAddressParameters::class.java)

        @JvmField
        val SERIALIZER: Serializer<BillingAddressParameters> = object : Serializer<BillingAddressParameters> {
            override fun serialize(modelObject: BillingAddressParameters): JSONObject {
                return try {
                    JSONObject().apply {
                        putOpt(FORMAT, modelObject.format)
                        putOpt(PHONE_NUMBER_REQUIRED, modelObject.isPhoneNumberRequired)
                    }
                } catch (e: JSONException) {
                    throw ModelSerializationException(BillingAddressParameters::class.java, e)
                }
            }

            override fun deserialize(jsonObject: JSONObject) = BillingAddressParameters(
                format = jsonObject.getStringOrNull(FORMAT),
                isPhoneNumberRequired = jsonObject.optBoolean(PHONE_NUMBER_REQUIRED),
            )
        }
    }
}
